packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -71,104 +105,297 @@ export function matches(element: Element, selector: AST): boolean {
+    // }
+    return Array.from(element.querySelectorAll(selector.content));
+  }
+  // Type of `complex` is used to express CSS combinators: ` `, `>`, `+`, `~`.
+  // The `branch` function describes the behavior per combinator.
+  if (selector.type === 'complex') {
+    return branch(element, selector);
+  }
+  // Type of `list` is sets of selector trees.
+  // We just join all the results.
+  // e.g. `p, span`
+  if (selector.type === 'list') {
+    const results: Element[] = [];
+    for (const item of selector.list) {
+      for (const result of querySelectorAll(element, item)) {
+        if (!results.includes(result)) {
+          results.push(result);
         }
       }
     }
-  } else if (selector.type === 'pseudo-class') {
-    // if (selector.name === 'upward') {
-    //   let n = Number(selector.argument);
-    //   console.log('upward', selector, n);
-    //   if (Number.isNaN(n) === false) {
-    //     if (n >= 1 && n < 256) {
-    //       let ancestor: Element | null = element;
-    //       while (ancestor !== null && n > 0) {
-    //         ancestor = ancestor.parentElement;
-    //         n -= 1;
-    //       }
-
-    //       if (ancestor !== null && n === 0) {
-    //         elements.push(element);
-    //       }
-    //     }
-    //   } else if (selector.argument !== undefined) {
-    //     const parent = element.parentElement;
-    //     if (parent !== null) {
-    //       const ancestor = parent.closest(selector.argument);
-    //       if (ancestor !== null) {
-    //         elements.push(ancestor);
-    //       }
-    //     }
-    //   }
-    // } else {
-    for (const subElement of element.querySelectorAll('*')) {
-      if (matches(subElement, selector) === true) {
-        elements.push(subElement);
+    return results;
+  }
+  // Type of `compound` is a set of consecutive selectors.
+  // They're in chained form like `p:has(span)` and works as logical AND.
+  if (selector.type === 'compound') {
+    const results: Element[] = [];
+    // Figure out the subjective element.
+    // If omitted, it will fallback to `pseudo-class` branch of `querySelectorAll`:
+    // - e.g. `:has(body)`, subjective is `html`.
+    // If given, it will run `querySelectorAll` on first selector:
+    // - e.g. `html:has(body)`, subjective is given.
+    const selectors = selector.compound.slice(1);
+    for (const subjective of querySelectorAll(element, selector.compound[0])) {
+      for (const result of traverse(subjective, selectors)) {
+        if (!results.includes(result)) {
+          results.push(result);
+        }
       }
     }
-    // }
+    return results;
   }
-
-  return elements;
+  // `querySelectorAll` is assumed as an entrypoint function and intended to run the first selector of the AST.
+  // Otherwise, we will run `traversals` directly by diciding the subjective element specifically.
+  // If `pseudo-class` is the first selector, it means `html` or the given element is omitted.
```

Comment:
The last line may be confusing: `the given element` need to be replaced with `the subjective element`.